ad_block_client.cc

```diff
@@ -1241,30 +1215,39 @@ bool AdBlockClient::parse(const char *input, bool preserveRules) {
       noFingerprintAntiDomainOnlyExceptionFilters) {
-    // Copy the old data in
-    memcpy(newFilters, filters, sizeof(Filter) * numFilters);
-    memcpy(newCosmeticFilters, cosmeticFilters,
-        sizeof(Filter) * numCosmeticFilters);
-    memcpy(newHtmlFilters, htmlFilters,
-        sizeof(Filter) * numHtmlFilters);
-    memcpy(newExceptionFilters, exceptionFilters,
-        sizeof(Filter) * numExceptionFilters);
-    memcpy(newNoFingerprintFilters, noFingerprintFilters,
-        sizeof(Filter) * (numNoFingerprintFilters));
-    memcpy(newNoFingerprintExceptionFilters, noFingerprintExceptionFilters,
-        sizeof(Filter) * (numNoFingerprintExceptionFilters));
-    memcpy(newNoFingerprintDomainOnlyFilters, noFingerprintDomainOnlyFilters,
-        sizeof(Filter) * (numNoFingerprintDomainOnlyFilters));
-    memcpy(newNoFingerprintAntiDomainOnlyFilters,
-        noFingerprintAntiDomainOnlyFilters,
-        sizeof(Filter) * (numNoFingerprintAntiDomainOnlyFilters));
-    memcpy(newNoFingerprintDomainOnlyExceptionFilters,
-        noFingerprintDomainOnlyExceptionFilters,
-        sizeof(Filter) * (numNoFingerprintDomainOnlyExceptionFilters));
-    memcpy(newNoFingerprintAntiDomainOnlyExceptionFilters,
-        noFingerprintAntiDomainOnlyExceptionFilters,
-        sizeof(Filter) * (numNoFingerprintAntiDomainOnlyExceptionFilters));
 
+    // Copy the old data in, we can't simply use memcpy here
+    // since filtres manages some pointers that get deleted.
+    for (int i = 0; i < numFilters; i++) {
+      newFilters[i].swapData(&(filters[i]));
+    }
+    for (int i = 0; i < numCosmeticFilters; i++) {
+      newCosmeticFilters[i].swapData(&(cosmeticFilters[i]));
+    }
+    for (int i = 0; i < numHtmlFilters; i++) {
+      newHtmlFilters[i].swapData(&(htmlFilters[i]));
+    }
+    for (int i = 0; i < numExceptionFilters; i++) {
+      newExceptionFilters[i].swapData(&(exceptionFilters[i]));
+    }
+    for (int i = 0; i < numNoFingerprintFilters; i++) {
+      newNoFingerprintFilters[i].swapData(&(noFingerprintFilters[i]));
+    }
+    for (int i = 0; i < numNoFingerprintExceptionFilters; i++) {
+      newNoFingerprintExceptionFilters[i].swapData(&(noFingerprintExceptionFilters[i]));
+    }
+    for (int i = 0; i < numNoFingerprintDomainOnlyFilters; i++) {
+      newNoFingerprintDomainOnlyFilters[i].swapData(&(noFingerprintDomainOnlyFilters[i]));
+    }
+    for (int i = 0; i < numNoFingerprintAntiDomainOnlyFilters; i++) {
+      newNoFingerprintAntiDomainOnlyFilters[i].swapData(&(noFingerprintAntiDomainOnlyFilters[i]));
+    }
+    for (int i = 0; i < numNoFingerprintDomainOnlyExceptionFilters; i++) {
+      newNoFingerprintDomainOnlyExceptionFilters[i].swapData(&(noFingerprintDomainOnlyExceptionFilters[i]));
+    }
+    for (int i = 0; i < numNoFingerprintAntiDomainOnlyExceptionFilters; i++) {
+      newNoFingerprintAntiDomainOnlyExceptionFilters[i].swapData(&(noFingerprintAntiDomainOnlyExceptionFilters[i]));
+    }
```

Comment:
Using the old way of simply memcpy over here no longer works because `Filter`'s store pointers now, and they manage their own memory. So when a Filter was destroyed, it would free the pointers twice.

Reply:
The perf here doesn't matter btw because this is only for parsing logic.   We use deserialized dat files in production.